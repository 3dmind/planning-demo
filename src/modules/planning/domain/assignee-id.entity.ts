import { Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';

export class AssigneeId extends Entity<null> {
  private constructor(id?: UniqueEntityId) {
    super(null, id);
  }

  get id(): UniqueEntityId {
    return this._id;
  }

  public static create(id?: UniqueEntityId): Result<AssigneeId> {
    return Result.ok<AssigneeId>(new AssigneeId(id));
  }

  public override toString(): string {
    return this._id.toString();
  }
}
