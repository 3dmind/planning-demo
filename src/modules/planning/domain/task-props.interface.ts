import { AssigneeId } from './assignee-id.entity';
import { Description } from './description.valueobject';
import { OwnerId } from './owner-id.entity';

export interface TaskProps {
  archived: boolean;
  archivedAt: Date;
  assigneeId: AssigneeId;
  createdAt: Date;
  description: Description;
  discarded: boolean;
  discardedAt: Date;
  editedAt: Date;
  ownerId: OwnerId;
  resumedAt: Date;
  tickedOff: boolean;
  tickedOffAt: Date;
}
