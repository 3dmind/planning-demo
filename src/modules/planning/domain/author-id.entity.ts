import { Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';

export class AuthorId extends Entity<null> {
  private constructor(id?: UniqueEntityId) {
    super(null, id);
  }

  get id(): UniqueEntityId {
    return this._id;
  }

  public static create(id?: UniqueEntityId): Result<AuthorId> {
    return Result.ok<AuthorId>(new AuthorId(id));
  }

  public override toString(): string {
    return this._id.toString();
  }
}
