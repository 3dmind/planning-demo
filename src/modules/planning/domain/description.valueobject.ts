import { Guard, Result } from '../../../shared/core';
import { ValueObject } from '../../../shared/domain';

interface DescriptionProps {
  value: string;
}

export class Description extends ValueObject<DescriptionProps> {
  public static readonly MIN_LENGTH = 2;
  public static readonly MAX_LENGTH = 250;

  private constructor(props: DescriptionProps) {
    super(props);
  }

  get value(): string {
    return this.props.value;
  }

  public static create(text: string): Result<Description> {
    const nullGuardResult = Guard.againstNullOrUndefined(text, 'text');

    if (!nullGuardResult.succeeded) {
      return Result.fail<Description>(nullGuardResult.message);
    }

    const minGuardResult = Guard.againstAtLeast(this.MIN_LENGTH, text);
    const maxGuardResult = Guard.againstAtMost(this.MAX_LENGTH, text);

    if (!minGuardResult.succeeded) {
      return Result.fail<Description>(minGuardResult.message);
    }

    if (!maxGuardResult.succeeded) {
      return Result.fail<Description>(maxGuardResult.message);
    }

    return Result.ok<Description>(new Description({ value: text }));
  }
}
