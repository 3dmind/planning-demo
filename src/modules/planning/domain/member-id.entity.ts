import { Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';

export class MemberId extends Entity<any> {
  constructor(id?: UniqueEntityId) {
    super(null, id);
  }

  get id(): UniqueEntityId {
    return this._id;
  }

  public static create(id?: UniqueEntityId): Result<MemberId> {
    return Result.ok<MemberId>(new MemberId(id));
  }

  public override toString(): string {
    return this._id.toString();
  }
}
