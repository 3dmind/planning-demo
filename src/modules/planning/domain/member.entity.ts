import { Guard, Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';
import { UserId } from '../../users/domain/user-id.entity';
import { AssigneeId } from './assignee-id.entity';
import { MemberId } from './member-id.entity';
import { MemberProps } from './member-props.interface';
import { OwnerId } from './owner-id.entity';

export class Member extends Entity<MemberProps> {
  private constructor(props: MemberProps, id?: UniqueEntityId) {
    super(props, id);
  }

  get memberId(): MemberId {
    return MemberId.create(this._id).getValue();
  }

  get ownerId(): OwnerId {
    return OwnerId.create(this._id).getValue();
  }

  get assigneeId(): AssigneeId {
    return AssigneeId.create(this._id).getValue();
  }

  get userId(): UserId {
    return this.props.userId;
  }

  get createdAt(): Date {
    return this.props.createdAt;
  }

  public static create(
    props: MemberProps,
    id?: UniqueEntityId,
  ): Result<Member> {
    const guardResult = Guard.againstNullOrUndefinedBulk([
      {
        argument: props.userId,
        argumentName: 'userId',
      },
    ]);

    if (!guardResult.succeeded) {
      return Result.fail<Member>(guardResult.message);
    }

    const member = new Member(
      {
        ...props,
        createdAt: props.createdAt ?? new Date(),
      },
      id,
    );
    return Result.ok<Member>(member);
  }
}
