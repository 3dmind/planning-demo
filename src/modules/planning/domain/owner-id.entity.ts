import { Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';

export class OwnerId extends Entity<null> {
  private constructor(id?: UniqueEntityId) {
    super(null, id);
  }

  get id(): UniqueEntityId {
    return this._id;
  }

  public static create(id?: UniqueEntityId): Result<OwnerId> {
    return Result.ok<OwnerId>(new OwnerId(id));
  }

  public toString(): string {
    return this._id.toString();
  }
}
