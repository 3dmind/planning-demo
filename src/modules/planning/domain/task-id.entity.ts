import { Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';

export class TaskId extends Entity<any> {
  private constructor(id?: UniqueEntityId) {
    super(null, id);
  }

  get id(): UniqueEntityId {
    return this._id;
  }

  public static create(id?: UniqueEntityId): Result<TaskId> {
    return Result.ok<TaskId>(new TaskId(id));
  }

  public override toString(): string {
    return this._id.toString();
  }
}
