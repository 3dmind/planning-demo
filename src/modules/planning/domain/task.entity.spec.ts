import * as faker from 'faker';
import * as uuid from 'uuid';
import { TaskEntityBuilder } from '../../../../test/builder/task-entity.builder';
import { UniqueEntityId } from '../../../shared/domain';
import { AssigneeId } from './assignee-id.entity';
import { Description } from './description.valueobject';
import { OwnerId } from './owner-id.entity';
import { TaskId } from './task-id.entity';
import { TaskProps } from './task-props.interface';
import { Task } from './task.entity';

jest.mock('uuid');

describe('Task', () => {
  beforeAll(() => {
    uuid.v4.mockReturnValue(faker.random.uuid());
  });

  afterAll(() => {
    uuid.mockRestore();
  });

  it('should guard "description" property', () => {
    expect.assertions(1);
    const props = { description: null } as TaskProps;

    const result = Task.create(props);

    expect(result.isFailure).toBe(true);
  });

  it('should guard "createdAt" property', () => {
    expect.assertions(1);
    const description = Description.create(faker.lorem.words(5)).getValue();
    const props = {
      description,
      createdAt: null,
    } as TaskProps;

    const result = Task.create(props);

    expect(result.isFailure).toBe(true);
  });

  it('should guard "tickedOff" property', () => {
    expect.assertions(1);
    const description = Description.create(faker.lorem.words(5)).getValue();
    const createdAt = new Date();
    const props = {
      description,
      createdAt,
      tickedOff: null,
    } as TaskProps;

    const result = Task.create(props);

    expect(result.isFailure).toBe(true);
  });

  it('should guard "archived" property', () => {
    expect.assertions(1);
    const description = Description.create(faker.lorem.words(5)).getValue();
    const createdAt = new Date();
    const tickedOff = false;
    const props = {
      description,
      createdAt,
      tickedOff,
      archived: null,
    } as TaskProps;

    const result = Task.create(props);

    expect(result.isFailure).toBe(true);
  });

  it('should guard "discarded" property', () => {
    expect.assertions(1);
    const description = Description.create(faker.lorem.words(5)).getValue();
    const createdAt = new Date();
    const tickedOff = false;
    const archived = false;
    const props = {
      description,
      createdAt,
      tickedOff,
      archived,
      discarded: null,
    } as TaskProps;

    const result = Task.create(props);

    expect(result.isFailure).toBe(true);
  });

  it('should guard "ownerId" property', () => {
    expect.assertions(1);
    const description = Description.create(faker.lorem.words(5)).getValue();
    const createdAt = new Date();
    const tickedOff = false;
    const archived = false;
    const discarded = false;
    const props = {
      description,
      createdAt,
      tickedOff,
      archived,
      discarded,
      ownerId: null,
    } as TaskProps;

    const result = Task.create(props);

    expect(result.isFailure).toBe(true);
  });

  it('should guard "assigneeId" property', () => {
    expect.assertions(1);
    const description = Description.create(faker.lorem.words(5)).getValue();
    const ownerId = OwnerId.create().getValue();
    const createdAt = new Date();
    const tickedOff = false;
    const archived = false;
    const discarded = false;
    const props = {
      description,
      createdAt,
      tickedOff,
      archived,
      discarded,
      ownerId,
      assigneeId: null,
    } as TaskProps;

    const result = Task.create(props);

    expect(result.isFailure).toBe(true);
  });

  it('should create task', () => {
    expect.assertions(5);
    const text = faker.lorem.words(5);
    const description = Description.create(text).getValue();
    const ownerId = OwnerId.create().getValue();
    const assigneeId = AssigneeId.create().getValue();
    const entityId = new UniqueEntityId();

    const taskResult = Task.create(
      {
        archived: false,
        archivedAt: null,
        assigneeId,
        createdAt: new Date(),
        description,
        discarded: false,
        discardedAt: null,
        editedAt: null,
        ownerId,
        resumedAt: null,
        tickedOff: false,
        tickedOffAt: null,
      },
      entityId,
    );
    const task = taskResult.getValue();

    expect(taskResult.isSuccess).toBe(true);
    expect(task.taskId).toBeInstanceOf(TaskId);
    expect(task.taskId.id.equals(entityId)).toBe(true);
    expect(task.ownerId.equals(ownerId)).toBe(true);
    expect(task.assigneeId.equals(assigneeId)).toBe(true);
  });

  it('should create snapshot', () => {
    expect.assertions(2);
    const task = new TaskEntityBuilder()
      .withDescription('Lorem ipsum')
      .withId(new UniqueEntityId('8bb2a83e-cf7e-46d7-a7b0-49f49d94c460'))
      .withOwnerId(
        OwnerId.create(
          new UniqueEntityId('74d21847-3298-4775-aaf8-942fed0f53e7'),
        ).getValue(),
      )
      .withAssigneeId(
        AssigneeId.create(
          new UniqueEntityId('27f0ca32-e888-4fcc-9908-b24bf152573d'),
        ).getValue(),
      )
      .withCreationDate(new Date(Date.parse('1977-01-01')))
      .build();

    const taskSnapshot = task.createSnapshot();

    expect(taskSnapshot).toMatchSnapshot();
    expect(Object.isFrozen(taskSnapshot)).toBe(true);
  });

  it('should note new task', () => {
    expect.assertions(1);
    const text = faker.lorem.words(5);
    const description = Description.create(text).getValue();
    const ownerId = OwnerId.create().getValue();
    const assigneeId = AssigneeId.create().getValue();

    const taskResult = Task.note(description, ownerId, assigneeId);

    expect(taskResult.isSuccess).toBe(true);
  });

  it('should tick off task', () => {
    const task = new TaskEntityBuilder().build();

    task.tickOff();

    expect(task.isTickedOff()).toBe(true);
  });

  it('should resume task', () => {
    expect.assertions(1);
    const task = new TaskEntityBuilder().makeTickedOff().build();

    task.resume();

    expect(task.isTickedOff()).toBe(false);
  });

  it('should archive task', () => {
    expect.assertions(1);
    const task = new TaskEntityBuilder().build();

    task.archive();

    expect(task.isArchived()).toBe(true);
  });

  it('should edit description', () => {
    expect.assertions(1);
    const text = faker.lorem.words(5);
    const task = new TaskEntityBuilder().withDescription(text).build();
    const newText = faker.lorem.words(5);
    const newDescription = Description.create(newText).getValue();

    task.edit(newDescription);

    expect(task.props.description.equals(newDescription)).toBe(true);
  });

  it('should discard task', () => {
    expect.assertions(1);
    const task = new TaskEntityBuilder().build();

    task.discard();

    expect(task.isDiscarded()).toBe(true);
  });
});
