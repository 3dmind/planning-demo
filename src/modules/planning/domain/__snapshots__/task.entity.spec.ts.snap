// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task should create snapshot 1`] = `
TaskSnapshot {
  "archivedAt": null,
  "assigneeId": AssigneeId {
    "_id": UniqueEntityId {
      "value": "27f0ca32-e888-4fcc-9908-b24bf152573d",
    },
    "props": null,
  },
  "createdAt": 1977-01-01T00:00:00.000Z,
  "description": Description {
    "props": Object {
      "value": "Lorem ipsum",
    },
  },
  "discardedAt": null,
  "editedAt": null,
  "isArchived": false,
  "isDiscarded": false,
  "isTickedOff": false,
  "ownerId": OwnerId {
    "_id": UniqueEntityId {
      "value": "74d21847-3298-4775-aaf8-942fed0f53e7",
    },
    "props": null,
  },
  "resumedAt": null,
  "taskId": TaskId {
    "_id": UniqueEntityId {
      "value": "8bb2a83e-cf7e-46d7-a7b0-49f49d94c460",
    },
    "props": null,
  },
  "tickedOffAt": null,
}
`;
